### ISSUE TITLE
Performance issue with Openssl 3.0 in multi threaded application when using d2i_x509

### ISSUE THREAD

**Issue by ramkumarpvs** [openssl/openssl#17950]
We are observing high CPU usage and performance issue with Openssl 3.0.1 on Linux when using Certificate extract API `d2i_X509` using default context. 
All our server processing threads are having high cpu usage and are spending most of their time in `CRYPTO_THREAD_read_lock, CRYPTO_THREAD_unlock and CRYPTO_THREAD_write_lock` similar to the one described in [#17064](https://github.com/openssl/openssl/issues/17064).
Each thread is showing CPU usage as high as 95-99%.

There are several code paths reaching the above crypto lock methods, we could capture stack traces and attached document for reference.

Stack traces for several threads:
[openssl3_0_d2iX509_filtered_stack_traces.txt]
(https://github.com/openssl/openssl/files/8336711/openssl3_0_d2iX509_filtered_stack_traces.txt)

Full stack trace for single thread:
[openssl3_0_d2iX509_full_stack_trace.txt](https://github.com/openssl/openssl/files/8336795/openssl3_0_d2iX509_full_stack_trace.txt)

From our application we use d2i_X509 which internally call ASN1_item_d2i_ex() across all threads and in most cases we see OSSL_DECODER_CTX_new_for_pkey() exists in the stack, would this locking issue be resolved through https://github.com/openssl/openssl/issues/17116 

We would like to know when this issue would be resolved. Let us know if any other information is required.

**Comment by ramkumarpvs** [comment]
@t8m , @paulidale , need your inputs on this issue.

**Comment by paulidale** [comment]
I suspect #17881 will make a difference once it is merged.  It's not clear if it will go to 3.0 as well at this point.

**Comment by paulidale** [comment]
As an interim fix, you could create a separate library context for each thread instead of sharing one.  It's not ideal but will prevent any contention.

**Comment by hlandau** [comment]
I'm seeing a substantial increase in performance relative to 3.0 when both my LIBCTX and LHASH patches are used.
```
# OpenSSL 3.0:
real    0m12.831s
user    0m29.056s
sys     0m6.494s

real    0m11.155s
user    0m21.563s
sys     0m4.564s

# OpenSSL master, plus my LIBCTX and LHASH patches:
real    0m9.399s
user    0m15.396s
sys     0m9.990s

real    0m9.645s
user    0m15.708s
sys     0m10.511s
```
So about 1.25x.

The following test program was used:
```c++
#include <openssl/x509.h>
#include <stdio.h>
#include <stdlib.h>
#include <vector>
#include <boost/thread/barrier.hpp>
#include <thread>

#define NUM_ITER 1000

int main(int argc, char **argv)
{
  FILE *f = fopen("x.der", "rb");
  if (!f) abort();

  fseek(f, 0, SEEK_END);
  long l = ftell(f);
  fseek(f, 0, SEEK_SET);
  unsigned char *der = (unsigned char *)malloc(l);
  if (fread(der, 1, l, f) < l) {
    fprintf(stderr, "read failure\n");
    return 1;
  }

  unsigned num_threads=10;
  std::vector<std::thread> threads;
  boost::barrier barrier{num_threads};
  for (size_t i=0; i<num_threads; ++i) {
    threads.emplace_back([der, l, &barrier]() {
      barrier.count_down_and_wait();
      for (size_t i=0; i<NUM_ITER; ++i) {
        X509 *cert = NULL;
        unsigned char *derp = der;
        X509 *c2 = d2i_X509(&cert, (const unsigned char **)&derp, l);
        if (c2 == NULL) {
          fprintf(stderr, "decode failure\n");
          abort();
        }
      }
    });
  }

  for (auto &t :threads)
    t.join();

  return 0;
}
```

Profiling after applying my LIBCTX and LHASH patches, ossl_sa_doall_arg does appear to be a hotspot. Worse though is that 24.4% of time is spent in __pthread_rwlock_unlock.

**Comment by paulidale** [comment]
`ossl_sa_doall_arg()` is meant to be pretty quick.  Any chance of a timing with a reduced size?  I.e. set `OPENSSL_SA_BLOCK_BITS` to something under 12 -- 4 to 6 ought to be enough to see a difference.

These timing programs you've put together should be stored somewhere -- we've a write some automated performance measurements task in the backlog.

**Comment by hlandau** [comment]
Scratch the above performance figures. I was working from a branch which didn't include my decoder optimizations.

Performance figures with 3.0:
```
real    0m14.138s
user    0m31.936s
sys     0m7.537s

real    0m14.316s
user    0m35.351s
sys     0m8.127s
```
Performance figures with both my merged and unmerged performance optimizations:
```
real    0m6.707s
user    0m10.702s
sys     0m7.116s

real    0m6.615s
user    0m10.624s
sys     0m6.841s
```
So about a 2.1x speedup.

`ossl_parse_property` is taking about 26% of the runtime according to callgrind. It's called from `ossl_decoder_from_algorithm`. On real hardware (perf(1)) 14% is taken up by __pthread_rwlock_unlock, but curiously the same isn't the case for locking. A large number of these lock acquisitions are due to `ossl_namemap_name2num_n`.
![Screenshot_2022-03-25_11-59-36](https://user-images.githubusercontent.com/2270010/160116926-8e905894-adfd-4c2f-a6cc-adb93635f206.png)

On a higher core count server (work, 24-core, 96-thread), 3.0:
```
$ ./d2i 10 1000

real    0m8.283s
user    0m38.707s
sys     0m9.608s
```
With my merged and unmerged performance optimizations:
```
$ ./d2i 10 1000
real    0m3.489s
user    0m6.763s
sys     0m5.141s
```
So about 2.4x.

Here are the runtimes for `t` threads, 1000 iterations where t \in [1,24] on the same server (2x Xeon 6248R):
```
===== Trying with 1 threads and 1000 iterations:

real    0m0.197s
user    0m0.188s
sys     0m0.009s

real    0m0.197s
user    0m0.189s
sys     0m0.008s
===== Trying with 2 threads and 1000 iterations:

real    0m0.411s
user    0m0.642s
sys     0m0.134s

real    0m0.396s
user    0m0.639s
sys     0m0.110s
===== Trying with 3 threads and 1000 iterations:

real    0m0.370s
user    0m0.833s
sys     0m0.149s

real    0m0.265s
user    0m0.684s
sys     0m0.042s
===== Trying with 4 threads and 1000 iterations:

real    0m1.281s
user    0m2.407s
sys     0m1.368s

real    0m1.249s
user    0m2.555s
sys     0m1.164s
===== Trying with 5 threads and 1000 iterations:

real    0m0.745s
user    0m2.182s
sys     0m0.459s

real    0m1.751s
user    0m3.591s
sys     0m1.956s
===== Trying with 6 threads and 1000 iterations:

real    0m2.603s
user    0m4.721s
sys     0m2.930s

real    0m2.282s
user    0m4.333s
sys     0m2.830s
===== Trying with 7 threads and 1000 iterations:

real    0m2.895s
user    0m5.590s
sys     0m3.452s

real    0m2.811s
user    0m5.475s
sys     0m3.345s
===== Trying with 8 threads and 1000 iterations:

real    0m3.152s
user    0m6.146s
sys     0m3.878s

real    0m3.300s
user    0m6.199s
sys     0m4.086s

===== Trying with 9 threads and 1000 iterations:

real    0m3.458s
user    0m6.650s
sys     0m4.675s

real    0m3.622s
user    0m6.949s
sys     0m4.684s
===== Trying with 10 threads and 1000 iterations:

real    0m3.452s
user    0m6.944s
sys     0m4.708s

real    0m4.141s
user    0m8.202s
sys     0m5.109s
===== Trying with 11 threads and 1000 iterations:

real    0m4.241s
user    0m8.219s
sys     0m5.647s

real    0m3.778s
user    0m7.652s
sys     0m5.351s
===== Trying with 12 threads and 1000 iterations:

real    0m4.419s
user    0m8.689s
sys     0m6.422s

real    0m5.224s
user    0m10.354s
sys     0m7.180s
===== Trying with 13 threads and 1000 iterations:

real    0m9.915s
user    0m19.741s
sys     0m13.324s

real    0m9.685s
user    0m19.756s
sys     0m12.999s
===== Trying with 14 threads and 1000 iterations:

real    0m11.300s
user    0m22.902s
sys     0m15.475s

real    0m12.326s
user    0m24.423s
sys     0m15.781s
===== Trying with 15 threads and 1000 iterations:

real    0m12.191s
user    0m25.496s
sys     0m16.437s

real    0m13.084s
user    0m26.302s
sys     0m16.434s
===== Trying with 16 threads and 1000 iterations:

real    0m13.556s
user    0m27.502s
sys     0m17.914s

real    0m13.923s
user    0m27.889s
sys     0m17.824s
===== Trying with 17 threads and 1000 iterations:

real    0m15.092s
user    0m29.989s
sys     0m19.136s

real    0m14.353s
user    0m28.390s
sys     0m20.375s
===== Trying with 18 threads and 1000 iterations:

real    0m15.635s
user    0m31.249s
sys     0m20.376s

real    0m15.244s
user    0m30.150s
sys     0m21.257s
===== Trying with 19 threads and 1000 iterations:

real    0m17.253s
user    0m34.116s
sys     0m21.738s

real    0m16.564s
user    0m32.855s
sys     0m21.559s
===== Trying with 20 threads and 1000 iterations:

real    0m17.175s
user    0m34.235s
sys     0m23.462s

real    0m17.312s
user    0m34.374s
sys     0m23.330s
===== Trying with 21 threads and 1000 iterations:

real    0m17.659s
user    0m35.992s
sys     0m23.697s

real    0m18.012s
user    0m36.031s
sys     0m24.359s
===== Trying with 22 threads and 1000 iterations:

real    0m19.225s
user    0m38.088s
sys     0m25.288s

real    0m18.963s
user    0m38.322s
sys     0m25.100s
===== Trying with 23 threads and 1000 iterations:

real    0m19.669s
user    0m40.152s
sys     0m27.238s

real    0m19.816s
user    0m39.663s
sys     0m26.847s
===== Trying with 24 threads and 1000 iterations:

real    0m20.470s
user    0m41.109s
sys     0m28.721s

real    0m22.129s
user    0m44.125s
sys     0m27.324s
```
As you can see, performance gets steadily worse as the thread count is increased.

This is related to #15199 but arguable whether it is a duplicate of it as it refers to the degradation of decode performance specifically in a highly parallel environment, rather than just low decode performance in general.

**Comment by mattcaswell** [comment]
> `ossl_parse_property` is taking about 26% of the runtime according to callgrind. It's called from `ossl_decoder_from_algorithm`.

This seems unexpected to me. If you are doing 1000 iterations I would expect `ossl_decoder_from_algorithm` to be called once per algorithm and then cached. While it might be a relatively expensive call, I would expect that overhead to disappear across 1000 iterations.

**Comment by t8m** [comment]
IMO that's because the decoding process is calling OSSL_DECODER_do_all_provided() which does not cache the results of ossl_decoder_from_algorithm(). If I understand the code correctly.

**Comment by mattcaswell** [comment]
> IMO that's because the decoding process is calling OSSL_DECODER_do_all_provided() which does not cache the results of ossl_decoder_from_algorithm(). If I understand the code correctly.

Hmm...are you sure about that? That would definitely be very expensive if we're not caching our fetches....but on a quick glance it looks to me like we are??

**Comment by mattcaswell** [comment]
I just ran a quick test using @hlandau reproducer to see how many times `ossl_decoder_from_algorithm` is called.

With a single thread and one iteration: 39
With a single thread and one thousand iterations: 39000

This is totally wrong if you ask me. This should all be cached. I would expect the second figure to be 39 also. No wonder performance is bad!

**Comment by mattcaswell** [comment]
I wonder if this explains the HPS numbers in #16791....obviously we call d2i_X509 as part of a TLS handshake.

**Comment by paulidale** [comment]
Only, it doesn't seem to.

**Comment by paulidale** [comment]
I suspect this might be fixed by #18151

**Comment by t8m** [comment]
Yes, this is a known issue. Could you please try OpenSSL 3.1 alpha 1? That should improve things at least partially.

**Comment by scaro-axway** [comment]
Thanks for your reply.
I built OpenSSL 3.1 alpha1 on the same system.
Performances are indeed much better than 3.0.7 (and still behind 1.1).
>    Version: OpenSSL 3.1.0-alpha1 1 Dec 2022
> Iterations: 100000
> real 3.59
> user 3.59
> sys 0.00

**Comment by t8m** [comment]
Yeah, unfortunately the decoder infrastructure is quite heavy. I am afraid it won't be improved besides the 3.1 results much any time soon.

**Comment by nhorman** [comment]
Marking as inactive, to be closed when 3.4 dev cycle ends, barring further input

**Comment by davidben** [comment]
What's the input that this is waiting on? Seems like this is already a performance report with a concrete and valuable benchmark.

**Comment by nhorman** [comment]
What I see in this issue is:
[A performance problem](https://github.com/openssl/openssl/issues/17950#issue-1178678738)
[A suggested fix](https://github.com/openssl/openssl/issues/17950#issuecomment-1076937751)
[Some commentary about other performance improvements that may or may not be related](https://github.com/openssl/openssl/issues/17950#issuecomment-1078827838)
[Some additional thoughts about performance](https://github.com/openssl/openssl/issues/17950#issuecomment-1079030921)
[Another suggested fix](https://github.com/openssl/openssl/issues/17950#issuecomment-1116937964)
[Another report of performance problems, which appears to be related](https://github.com/openssl/openssl/issues/17950#issuecomment-1334913053)
[Another suggestion for a fix](https://github.com/openssl/openssl/issues/17950#issuecomment-1334913053)
[A response indicating that performance is improved, but still less than 1.1](https://github.com/openssl/openssl/issues/17950#issuecomment-1334913053)
[A comment indicating that decoder performance is not likely to reach that of 1.1](https://github.com/openssl/openssl/issues/17950#issuecomment-1334913053)

Followed by 2 years of silence

So my question is valid, we need some level of input to determine what, if anything to do next with this issue:
* from a team member here, indicating a next step to take to improve performance
* from the reporter, indicating that performance in this area is not yet sufficient for their needs, and what their testing shows is a problem in light of the improvements that have already been made
* a statement from a community member indicating that they're interested in pursuing this in light of our limited capacity to address this particular issue.


Barring any of this, Its seems all that can currently be done, has been done here, and this issue can be closed.

**Comment by davidben** [comment]
Have you all run this through a profiler and investigated things, or is this just a generic guess?

For the private key parser, I recall the refrain was that the performance regression was a necessary cost of the more flexible decoder architecture. Then I investigated and found that actually the decoder architecture was staggeringly inefficient for reasons that had nothing to do with the architecture's state design goals. It was simply designed wrong. See https://github.com/openssl/openssl/issues/15199#issuecomment-1631038583

I am unlikely to have time to investigate this for you all, as this isn't my day job and I'm a bit hosed, but it seems to me:

* Certificate decoding performance would be quite an important use case for OpenSSL users
* The bug already comes with an easy to run benchmark to test against 
* Much more investigation is warranted to determine if this is the same bad architecture, an unrelated issue, or a concrete, explainable cause that is a justifiable regression

Thus this triage decision seems bizarre. Of course, you all can run your project as you see fit, but I find this a bit baffling.

**Comment by nhorman** [comment]
so, I took a look at this today.

using the following test program, a variant of those found in our perftools lib:
```
/*
 * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
 *
 * Licensed under the Apache License 2.0 (the "License").  You may not use
 * this file except in compliance with the License.  You can obtain a copy
 * in the file LICENSE in the source distribution or at
 * https://www.openssl.org/source/license.html
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#ifndef _WIN32
# include <libgen.h>
# include <unistd.h>
#else
# include <windows.h>
# include "perflib/getopt.h"
# include "perflib/basename.h"
#endif	/* _WIN32 */
#include <openssl/ssl.h>
#include <openssl/crypto.h>
#include "perflib/perflib.h"

#define NUM_CALLS_PER_TEST         1000000

static const unsigned char derstring[] = {
0x30,0x82,0x03,0x25,0x30,0x82,0x02,0x0d,0xa0,0x03,0x02,0x01,0x02,0x02,0x01,0x02,
0x30,0x0d,0x06,0x09,0x2a,0x86,0x48,0x86,0xf7,0x0d,0x01,0x01,0x0b,0x05,0x00,0x30,
0x12,0x31,0x10,0x30,0x0e,0x06,0x03,0x55,0x04,0x03,0x0c,0x07,0x52,0x6f,0x6f,0x74,
0x20,0x43,0x41,0x30,0x20,0x17,0x0d,0x31,0x36,0x30,0x31,0x31,0x34,0x32,0x32,0x32,
0x39,0x34,0x36,0x5a,0x18,0x0f,0x32,0x31,0x31,0x36,0x30,0x31,0x31,0x35,0x32,0x32,
0x32,0x39,0x34,0x36,0x5a,0x30,0x19,0x31,0x17,0x30,0x15,0x06,0x03,0x55,0x04,0x03,
0x0c,0x0e,0x73,0x65,0x72,0x76,0x65,0x72,0x2e,0x65,0x78,0x61,0x6d,0x70,0x6c,0x65,
0x30,0x82,0x01,0x22,0x30,0x0d,0x06,0x09,0x2a,0x86,0x48,0x86,0xf7,0x0d,0x01,0x01,
0x01,0x05,0x00,0x03,0x82,0x01,0x0f,0x00,0x30,0x82,0x01,0x0a,0x02,0x82,0x01,0x01,
0x00,0xd5,0x5d,0x60,0x6a,0xdf,0xfc,0x61,0xee,0x48,0xaa,0x8c,0x11,0x48,0x43,0xa5,
0x6d,0xb6,0x52,0x5d,0xaa,0x98,0x49,0xb1,0x61,0x92,0x35,0xb1,0xfc,0x3a,0x04,0x25,
0x0c,0x6d,0x79,0xff,0xb4,0xd5,0xc9,0xe9,0x5c,0x1c,0x3b,0xe0,0xab,0xb3,0xb8,0x7d,
0xa3,0xde,0x6d,0xbd,0xe0,0xdd,0xd7,0x5a,0xbf,0x14,0x47,0x11,0x42,0x5e,0xa6,0x82,
0xd0,0x61,0xc1,0x7f,0xdd,0x13,0x46,0xe6,0x09,0x85,0x07,0x0e,0xf2,0xd4,0xfc,0x1a,
0x64,0xd2,0x0a,0xad,0x20,0xab,0x20,0x6b,0x96,0xf0,0xad,0xcc,0xc4,0x19,0x53,0x55,
0xdc,0x01,0x1d,0xa4,0xb3,0xef,0x8a,0xb4,0x49,0x53,0x5d,0x8a,0x05,0x1c,0xf1,0xdc,
0xe1,0x44,0xbf,0xc5,0xd7,0xe2,0x77,0x19,0x57,0x5c,0x97,0x0b,0x75,0xee,0x88,0x43,
0x71,0x0f,0xca,0x6c,0xc1,0xb4,0xb2,0x50,0xa7,0x77,0x46,0x6c,0x58,0x0f,0x11,0xbf,
0xf1,0x76,0x24,0x5a,0xae,0x39,0x42,0xb7,0x51,0x67,0x29,0xe1,0xd0,0x55,0x30,0x6f,
0x17,0xe4,0x91,0xea,0xad,0xf8,0x28,0xc2,0x43,0x6f,0xa2,0x64,0xa9,0xfb,0x9d,0x98,
0x92,0x62,0x48,0x3e,0xeb,0x0d,0x4f,0x82,0x4a,0x8a,0xff,0x3f,0x72,0xee,0x96,0xb5,
0xae,0xa1,0xc1,0x98,0xba,0xef,0x7d,0x90,0x75,0x6d,0xff,0x5a,0x52,0x9e,0xab,0xf5,
0xc0,0x7e,0xd0,0x87,0x43,0xdb,0x85,0x07,0x07,0x0f,0x7d,0x38,0x7a,0xfd,0xd1,0xd3,
0xee,0x65,0x1d,0xd3,0xea,0x39,0x6a,0x87,0x37,0xee,0x4a,0xd3,0xe0,0x0d,0x6e,0xf5,
0x70,0xac,0xc2,0xbd,0xf1,0x6e,0xf3,0x92,0x95,0x5e,0xa9,0xf0,0xa1,0x65,0x95,0x93,
0x8d,0x02,0x03,0x01,0x00,0x01,0xa3,0x7d,0x30,0x7b,0x30,0x1d,0x06,0x03,0x55,0x1d,
0x0e,0x04,0x16,0x04,0x14,0xc0,0xe7,0x84,0xbf,0xe8,0x59,0x27,0x33,0x10,0xb0,0x52,
0x4f,0x51,0x52,0x2f,0x06,0xd6,0xc0,0x7a,0xcd,0x30,0x1f,0x06,0x03,0x55,0x1d,0x23,
0x04,0x18,0x30,0x16,0x80,0x14,0x70,0x7f,0x2e,0xae,0x83,0x68,0x59,0x98,0x04,0x23,
0x2a,0xcd,0xeb,0x3e,0x17,0xcd,0x24,0xdd,0x01,0x49,0x30,0x09,0x06,0x03,0x55,0x1d,
0x13,0x04,0x02,0x30,0x00,0x30,0x13,0x06,0x03,0x55,0x1d,0x25,0x04,0x0c,0x30,0x0a,
0x06,0x08,0x2b,0x06,0x01,0x05,0x05,0x07,0x03,0x01,0x30,0x19,0x06,0x03,0x55,0x1d,
0x11,0x04,0x12,0x30,0x10,0x82,0x0e,0x73,0x65,0x72,0x76,0x65,0x72,0x2e,0x65,0x78,
0x61,0x6d,0x70,0x6c,0x65,0x30,0x0d,0x06,0x09,0x2a,0x86,0x48,0x86,0xf7,0x0d,0x01,
0x01,0x0b,0x05,0x00,0x03,0x82,0x01,0x01,0x00,0x7b,0xd3,0x04,0x43,0x75,0x8a,0x0f,
0x11,0xae,0xc4,0xfb,0xd7,0xa1,0xa2,0x9e,0xfe,0x20,0x18,0xd5,0xf4,0x2f,0x31,0x88,
0x46,0xb6,0x75,0x8c,0xee,0xe5,0x9b,0x97,0xa6,0xb9,0xa3,0xcd,0x60,0x9a,0x46,0xc3,
0x48,0x97,0xe5,0x97,0x68,0xf7,0x5a,0x86,0x35,0x73,0xd9,0x69,0x9e,0xf9,0x5f,0x74,
0xb9,0xe6,0x94,0x13,0x01,0xcb,0x6a,0xdc,0xe3,0xc4,0x04,0xe9,0x65,0xda,0x9c,0xa4,
0x8b,0x28,0xf3,0xf9,0x9a,0x7f,0xbf,0x97,0x1f,0x45,0x92,0xe5,0x05,0xb1,0x56,0xe6,
0x0b,0xf6,0x47,0xde,0x1e,0x89,0xb6,0x2b,0xe1,0x4d,0xdf,0x4a,0x7e,0x01,0xd3,0x23,
0xdc,0x97,0x8c,0x47,0xfe,0x5f,0xc7,0xcc,0x98,0x46,0x0e,0xc4,0x83,0x5b,0xca,0x8a,
0xf1,0x52,0x09,0xbe,0x6b,0xec,0x3f,0x09,0x8b,0xd0,0x93,0x02,0xbf,0xe1,0x51,0xe7,
0xd1,0x7e,0x34,0x56,0x19,0x74,0xd0,0xff,0x28,0x25,0xde,0xb7,0x9f,0x56,0x52,0x91,
0x7d,0x20,0x29,0x85,0x0a,0x80,0x44,0x5f,0x71,0x32,0x25,0x71,0x0f,0xc2,0x16,0xe2,
0x5f,0x6b,0x1d,0x3f,0x32,0x5b,0x0a,0x3c,0x74,0x1c,0xb9,0x62,0xf1,0xed,0x07,0x50,
0xa3,0x6d,0xb4,0xb4,0x31,0x0a,0xc0,0x53,0x44,0x6a,0x3a,0x88,0x84,0x8b,0x2d,0xa9,
0xb0,0x37,0x8e,0xe6,0x18,0x36,0xbd,0x9a,0x20,0x40,0x0f,0x01,0x92,0x8b,0x3d,0xaa,
0x61,0xe7,0xae,0x2c,0xed,0x36,0xcd,0x3a,0x07,0x86,0x74,0x3a,0x29,0xb3,0xd7,0x3a,
0xb4,0x00,0xa9,0xc2,0xf5,0x92,0x78,0x0e,0xe2,0x0f,0xa3,0xfe,0xbb,0xbe,0xe0,0x06,
0x53,0x84,0x59,0x1d,0x90,0x69,0xe5,0xb6,0xf9
};

int err = 0;
static SSL_CTX *ctx;

size_t num_calls;
static int threadcount;
static OSSL_TIME *times = NULL;

void do_d2i(size_t num)
{
    size_t i;
    X509 *cert = NULL;
    OSSL_TIME start, end;
    const unsigned char *dstring;

    start = ossl_time_now();

    for (i = 0; i < num_calls / threadcount; i++) {
        dstring = derstring;
        cert = d2i_X509(&cert, (const unsigned char **)&dstring, 809);    
        if (cert == NULL) {
            fprintf(stderr, "Failed to convert certificate\n");
            break;
        }
        X509_free(cert);
        cert = NULL;
    }

    end = ossl_time_now();
    times[num] = ossl_time_subtract(end, start);
}

int main(int argc, char *argv[])
{
    OSSL_TIME duration;
    OSSL_TIME ttime;
    double avcalltime;
    int terse = 0;
    int rc = EXIT_FAILURE;
    size_t i;
    int opt;

    while ((opt = getopt(argc, argv, "t")) != -1) {
        switch (opt) {
        case 't':
            terse = 1;
            break;
        default:
            printf("Usage: %s [-t] threadcount\n", basename(argv[0]));
            printf("-t - terse output\n");
            return EXIT_FAILURE;
        }
    }

    if (argv[optind] == NULL) {
        printf("threadcount is missing\n");
        return EXIT_FAILURE;
    }
    threadcount = atoi(argv[optind]);
    if (threadcount < 1) {
        printf("threadcount must be > 0\n");
        return EXIT_FAILURE;
    }
    num_calls = NUM_CALLS_PER_TEST;
    if (NUM_CALLS_PER_TEST % threadcount > 0) /* round up */
        num_calls += threadcount - NUM_CALLS_PER_TEST % threadcount;

    times = OPENSSL_malloc(sizeof(OSSL_TIME) * threadcount);
    if (times == NULL) {
        printf("Failed to create times array\n");
        return EXIT_FAILURE;
    }

    if (!perflib_run_multi_thread_test(do_d2i, threadcount, &duration)) {
        printf("Failed to run the test\n");
        goto out;
    }

    if (err) {
        printf("Error during test\n");
        goto out;
    }

    ttime = times[0];
    for (i = 1; i < threadcount; i++)
        ttime = ossl_time_add(ttime, times[i]);

    avcalltime = ((double)ossl_time2ticks(ttime) / num_calls) / (double)OSSL_TIME_US;

    if (terse)
        printf("%lf\n", avcalltime);
    else
        printf("Average time per d2i decode call: %lfus\n",
               avcalltime);
    rc = EXIT_SUCCESS;
out:
    OPENSSL_free(times);
    return rc;
}
```


Running this on 10 threads on my system for a total of 100000 iterations I get the following:
against the master branch:
```
./d2i_master 10
Average time per d2i decode call: 10.083908us
```

against the 1.1.1w branch:
```
./d2i_11 10
Average time per d2i decode call: 5.850010us
```

So I see a slowdown of almost 100%.

Running it under perf shows that a significant amount of time is spent in fnv1a_hash, as a derivative of lots of calls to ossl_namemap_name2num and its callers:
![Screenshot from 2024-10-30 13-34-02](https://github.com/user-attachments/assets/4746eda6-3270-492e-b9c2-401c6dba8d62)


1.1.1 doesn't do decoding in the same way that 1.1.1 does of course, so this is work we're doing in 3.0 that we never had to do in 1.1.1.  this is the decoder impact that @t8m noted [here](https://github.com/openssl/openssl/issues/17950#issuecomment-1335108227)

Not sure what to do about that yet, but I'm looking at it

**Comment by r4ntix** [comment]
Is there an update? we are also seeing this issue(high cpu usage) in our production environment.

**Comment by jedisct1** [comment]
> Is there an update? we are also seeing this issue(high cpu usage) in our production environment.

Do you see any?
